includes/functions.php

```diff
@@ -2310,4 +2374,901 @@ function tsml_date_localised($format, $timestamp = null)
+                $message_lines[] = $message;
+                //add changed record id to list of db records to be removed
+                $db_ids_to_delete[] = $db_meeting_id;
+            } else {
+                $message = __("<tr style='color:green;'><td>Add New</td><td >$meeting_name_linked</td><td>$day_of_week</td><td>$meeting_time_formatted</td><td>$last_update</td></tr>", '12-step-meeting-list');
+                $message_lines[] = $message;
+            }
+        }
+    }
+
+    //update removal array with only those meetings in local database which are not matched with the import feed
+    foreach ($db_meetings as $db_meeting) {
+
+        list($day_of_week, $dow_number) = tsml_get_day_of_week_info($db_meeting['day'], $tsml_days);
+        $meeting_slug = $db_meeting['slug'];
+
+        $is_matched = in_array($meeting_slug, $feed_slugs);
+
+        // Check if slug has been modified on import by removing an appended suffix and test for match again
+        if (!$is_matched) {
+            for ($x = 0; $x <= 10; $x++) {
+                if (str_contains($meeting_slug, '-' . $x)) {
+                    $meeting_slug = str_replace('-' . $x, '', $meeting_slug);
+                    break;
+                }
+            }
+            $is_matched = in_array($meeting_slug, $feed_slugs);
+        }
+
+        if (!$is_matched) {
+            $meeting_update_date = date(get_option('date_format') . ' ' . get_option('time_format'), $data_source_last_update);
+            $meeting_name = $db_meeting['name'];
+            $permalink = get_permalink($db_meeting['id']);
+            $meeting_name_linked = '<a href=' . $permalink . '>' . $meeting_name . '</a>';
+			$meeting_time = $db_meeting['time'];
+			$meeting_time_formatted = date(get_option('time_format'), strtotime($meeting_time));
+
+            $message = __("<tr style='color:red;'><td>Remove</td><td >$meeting_name_linked</td><td>$day_of_week</td><td><td>$meeting_time_formatted</td>* $meeting_update_date</td></tr>", '12-step-meeting-list');
+            $message_lines[] = $message;
+            $db_ids_to_delete[] = $db_meeting['id'];
+        }
+    }
+
+    return $meetings_updated;
+}
+
+//function:	return boolean indicating whether or not the duplicate named meetings found in a data source db recordset are actual matches based on day, time, and region
+//used:		here (called from tsml_get_import_changes_only)
+function tsml_do_critical_fields_match($meeting, $db_meeting) {
+
+	if (intval($meeting['day']) !== intval($db_meeting['day'])) {
+		return false;
+	}
+	if ( $meeting['time'] !==  $db_meeting['time']) {
+		return false;
+	}
+	if ( $meeting['region'] !==  $db_meeting['region']) {
+		return false;
+	}
+	return true;
+}
+
+//function:	return boolean indicating whether or not the matched database and import records are in sync
+//used:		here (called from tsml_get_import_changes_only)
+function tsml_verify_identical_records($db_meeting, &$import_meeting, $data_source_last_update)
```

Comment:
this function has to use a repeat loop - it's far too long - 668 lines!

Reply:
It's doing a lot when making a differential check.  It's already in the Input Feed loop where it matches up the incoming record with the database record to determine if there is a change. What are you proposing  to loop?

Reply:
If you study the tsml_verify_identical_records code, you may notice that it's not a straight array to array comparison having the same key/value pairs (which I'm sure you're aware of). Depending on the $meeting array key we have to figure out the correct db_meeting field to compare with. Once we detect a change, the record is marked as changed and the loop is terminated. Note that the "updated" field is the very last field to be checked.

Reply:
sorry, i should have been more explicit in my feedback. a lot of the length in the PR looks like it comes from comparing very similar text fields. 

you could try something like the following?

```
$text_fields = [
  'website' => 'Website',
  'group' => 'Group',
  'group_notes' => 'Group Notes',
  'other_field' => 'Other FIeld',
];

foreach ($text_fields as $key => $name) {
    if (!empty($db_meeting[$key]) && !empty($import_meeting[$key)) {
         $db_value = tsml_remove_special_char($db_meeting[$key]);
         $import_value = tsml_remove_special_char($import_meeting[$key]);

         if ($db_value !== $import_value) {

             if ($test_mode) {
                 echo $name . ' --→ ' . $import_meeting['name'] . '<br>';
                 echo $db_value . '  <br>';
                 echo $import_value . ' <br><br>';
             }

             return false;
         }
     }
}
```

this could probably be carried further by using the existing `$tsml_export_columns` variable and creating conditionals for non-standard fields:

```
if ($key === 'types') {
  $db_value = serialize($db_meeting[$key]);
  $import_value = serialize($import_meeting[$key]);
}
```